70_FLIPFluids/app_resources/compute/pressureSolver.comp.hlsl

```diff
@@ -0,0 +1,120 @@
+[[vk::binding(b_psPresOutBuffer, s_ps)]] RWStructuredBuffer<float> pressureOutBuffer;
+
+[numthreads(WorkgroupSize, 1, 1)]
+void calculateNegativeDivergence(uint32_t3 ID : SV_DispatchThreadID)
+{
+    uint cid = ID.x;
+    int3 cellIdx = flatIdxToCellIdx(cid, gridData.gridSize);
+
+    float3 param = (float3)gridData.gridInvCellSize;
+    float3 velocity = velocityFieldBuffer[cellIdx].xyz;
+
+    float divergence = 0;
+    if (isFluidCell(getCellMaterial(cellMaterialBuffer[cid])))
+    {
+        int3 cell_xn = cellIdx + int3(1, 0, 0);
+        uint cid_xn = cellIdxToFlatIdx(cell_xn, gridData.gridSize);
+        divergence += param.x * ((cell_xn.x < gridData.gridSize.x ? velocityFieldBuffer[cell_xn].x : 0.0f) - velocity.x);
+
+        int3 cell_yn = cellIdx + int3(0, 1, 0);
+        uint cid_yn = cellIdxToFlatIdx(cell_yn, gridData.gridSize);
+        divergence += param.y * ((cell_yn.y < gridData.gridSize.y ? velocityFieldBuffer[cell_yn].y : 0.0f) - velocity.y);
+
+        int3 cell_zn = cellIdx + int3(0, 0, 1);
+        uint cid_zn = cellIdxToFlatIdx(cell_zn, gridData.gridSize);
+        divergence += param.z * ((cell_zn.z < gridData.gridSize.z ? velocityFieldBuffer[cell_zn].z : 0.0f) - velocity.z);
+    }
+
+    divergenceBuffer[cid] = divergence;
+}
+
+[numthreads(WorkgroupSize, 1, 1)]
+void solvePressureSystem(uint32_t3 ID : SV_DispatchThreadID)
+{
+    uint cid = ID.x;
+    int3 cellIdx = flatIdxToCellIdx(cid, gridData.gridSize);
+
+    float pressure = 0;
+
+    uint cellMaterial = cellMaterialBuffer[cid];
+
+    if (isFluidCell(getCellMaterial(cellMaterial)))
+    {
+        uint cid_xp = cellIdxToFlatIdx(cellIdx + int3(-1, 0, 0), gridData.gridSize);
+        cid_xp = isSolidCell(getXPrevMaterial(cellMaterial)) ? cid : cid_xp;
+        uint cid_xn = cellIdxToFlatIdx(cellIdx + int3(1, 0, 0), gridData.gridSize);
+        cid_xn = isSolidCell(getXNextMaterial(cellMaterial)) ? cid : cid_xn;
+        pressure += params.coeff1.x * (pressureInBuffer[cid_xp] + pressureInBuffer[cid_xn]);
+
+        uint cid_yp = cellIdxToFlatIdx(cellIdx + int3(0, -1, 0), gridData.gridSize);
+        cid_yp = isSolidCell(getYPrevMaterial(cellMaterial)) ? cid : cid_yp;
+        uint cid_yn = cellIdxToFlatIdx(cellIdx + int3(0, 1, 0), gridData.gridSize);
+        cid_yn = isSolidCell(getYNextMaterial(cellMaterial)) ? cid : cid_yn;
+        pressure += params.coeff1.y * (pressureInBuffer[cid_yp] + pressureInBuffer[cid_yn]);
+
+        uint cid_zp = cellIdxToFlatIdx(cellIdx + int3(0, 0, -1), gridData.gridSize);
+        cid_zp = isSolidCell(getZPrevMaterial(cellMaterial)) ? cid : cid_zp;
+        uint cid_zn = cellIdxToFlatIdx(cellIdx + int3(0, 0, 1), gridData.gridSize);
+        cid_zn = isSolidCell(getZNextMaterial(cellMaterial)) ? cid : cid_zn;
+        pressure += params.coeff1.z * (pressureInBuffer[cid_zp] + pressureInBuffer[cid_zn]);
+
+        pressure += params.coeff1.w * divergenceBuffer[cid];
+    }
+
+    pressureOutBuffer[cid] = pressure;
+}
```

Comment:
a 8x8x8 dispatch has 512 threads.

Recent Nvidia GPUs can have 2,3 or 4 of those resident on an SM at the same time.

Vulkan requires implementations supports at least 16kb of shared memory per workgroup, but in practice all desktop GPUs let you use 32kb.

It just so happens that the old NV GPU arch (Pascal) that can have max 2 workgroups co-resident has at least 64kb of smem, so enough to support 2 workgroups using 32kb each, and newer GPUs have even more.

You are accessing:
- divergence scalar
- pressure scalar
- solid testing 

32kb is enough to store 5461 cells of this data if you use float16_t, and 3276 if you use float32_t for the scalars.

These mean 17^3 or 14^3 grids are possible to keep in shared memory. *

Meaning you can do (17-8)/2=4 or  (14-8)/2=3 iterations of pressure solving in a single dispatch.

* How is it possible to do only have one copy of your buffers without ping-ponging? Red-Black Ordering!
https://www.youtube.com/watch?app=desktop&v=giTZ89q-Bpk